introduce a preserve command line switch in debugfs 'rdump' command

qkaiser opened the pull request:
`debugfs` assumed the user wanted to preserve permissions and ownership when dumping a filesystem directory recursively with `rdump`. This is in opposition with the way the `dump/dump_inode` command has been designed, since it expose a `-p` command line switch to allow the end users to explicitly opt-in for permission and ownership preservation.
    
The inability to explicitly ask for permission and ownership preservation would get rdump to default to preservation, which is a problem when faced with filesystems having directories with the **read** flag set but not the **execute** flag set, since it would only allow to enumerate the directory content, but not see the inode details. Therefore getting debugfs in all kinds of issues trying to set ownership and permissions of files it can't see. 

This fix introduce a 'preserve' (`-p`) flag in rdump command so that users can explicitly opt-in for it, and debugfs will default to a safer way of operation (no preserve).

---

Note: I also fixed some code to get it to build with modern clang, and formatted the code with `clang-format` because the code is unreadable otherwise.

qkaiser committed:
chore: fix modern clang errors.

Unused variable ext_written was reported by -Wunused-but-set-variable,
so it's now guarded by the same preprocessor guards that protects its
actual usage.

Function without prototypes are note allowed in C23.

vlaci commented:
We should decide if we want to reformat the code or not. If there are ever new upstream changes in this tool, it will be hell to rebase to. If we are to take the risk, we should first reformat the code, and implement the new functionality on top, so that functional changes are not mixed with formatting ones on the same base.
Personally, I am for to keep the code style as is.

qkaiser commented:
@vlaci this is an excellent point, I'll rework the MR without formatting stuff so we're safe if we have to rebase on upstream in the future.

qkaiser committed:
debugfs: expose a preserve command line switch in rdump command

rdump assumed the user wanted to preserve permissions and ownership when
dumping a filesystem directory recursively with 'rdump'. This is in
opposition with the way the 'dump' or 'dump_inode' command has been
designed, since it expose a '-p' command line switch to allow the end
users to explicitly opt-in for permission and ownership preservation.

The inability to explicitly ask for permission and ownership
preservation would get rdump to default to preservation, which is a
problem when faced with filesystems having directories with the read
flag but not the execute flag, since it would only allow to enumerate
the directory content, but not see the inode details. Therefore getting
debugfs in all kinds of issues trying to set ownership and permissions
of files it can't see.

This fix introduce a 'preserve' ('-p') flag in rdump command so that
users can explicitly opt-in for it, and debugfs will default to a safer
way of operation (no preserve).

qkaiser commented:
This MR now only contains changes to the code itself and no reformatting. Code changes stick to tabs rather than spaces since it's what was used originally.

orosam commented on debugfs/dump.c:
```diff
@@ -348,19 +360,36 @@ static int rdump_dirent(struct ext2_dir_entry *dirent,
+				preserve++;
+				break;
+			default:
+			print_usage:
```
This print_usage block should be moved to the end of the function after the final return, and goto-ed from here instead.

orosam commented on debugfs/dump.c:
```diff
@@ -348,19 +360,36 @@ static int rdump_dirent(struct ext2_dir_entry *dirent,
+			print_usage:
+				com_err(argv[0], 0,
+						"Usage: rdump [-p] "
+						"[<fs_directory_1> <fs_directory_2> ...] <output_file>");
```
Isn't this rather something like 
`<fs_directory>... <output_directory>` ? This would mean that at least one fs_directory and exactly one output directory is necessary.

orosam commented on debugfs/dump.c:
```diff
@@ -348,19 +360,36 @@ static int rdump_dirent(struct ext2_dir_entry *dirent,
+				com_err(argv[0], 0,
+						"Usage: rdump [-p] "
+						"[<fs_directory_1> <fs_directory_2> ...] <output_file>");
+				exit(1);
```
We can use `ss_set_exit_status(sci_idx, 1)` here as well.

orosam commented on debugfs/dump.c:
```diff
@@ -262,10 +262,15 @@ static void rdump_symlink(ext2_ino_t ino, struct ext2_inode *inode,
 	free(buf);
 }
 
+typedef struct rdump_dirent_private {
```
An anonymous struct would do (it's already used elsewhere):

```
typedef struct {
...
} rdump_dirent_private;
```

and replace `rdump_dirent_private_t` with `rdump_dirent_private` below.

orosam commented on debugfs/dump.c:
```diff
@@ -315,14 +320,22 @@ static int rdump_inode(ext2_ino_t ino, struct ext2_inode *inode,
 
+		rdump_dirent_private_t* entry = malloc(sizeof(rdump_dirent_private_t));
+		if (entry == NULL) {
+			com_err("rdump", errno, "while allocating entry for %s", fullname);
```
And do a `goto errout;`, otherwise it's a null pointer deref two lines below.

orosam commented on debugfs/dump.c:
```diff
@@ -315,14 +320,22 @@ static int rdump_inode(ext2_ino_t ino, struct ext2_inode *inode,
-
-		fix_perms("rdump", inode, -1, fullname);
+		if (preserve)
+		    fix_perms("rdump", inode, -1, fullname);
```
#trivial: use a single tab for indent.

orosam commented on debugfs/dump.c:
```diff
@@ -338,7 +351,6 @@ static int rdump_dirent(struct ext2_dir_entry *dirent,
 {
 	char name[EXT2_NAME_LEN + 1];
 	int thislen;
-	const char *dumproot = private;
```
It would be more readable to cast `private` here up front:
`rdump_dirent_private *dirent_private = (rdump_dirent_private *) private;`

And then use `dirent_private` in the `rdump_inode()` call below.

orosam commented on debugfs/dump.c:
```diff
@@ -392,9 +418,14 @@ void do_rdump(int argc, char **argv, int sci_idx,
+		if (rdump_inode(ino, &inode, basename, dest_dir, preserve) != 0)
 			ss_set_exit_status(sci_idx, 1);
 	}
+	print_usage:
+		com_err(argv[0], 0,
+				"Usage: rdump [-p] "
+				"<fs_directory>... <output_file>");
+		ss_set_exit_status(sci_idx, 1);
```
A `return` should be added before the label.
Also, two minor changes: indent should be decreased one level, and output_file changed to output_directory.

```suggestion
	return;

print_usage:
	com_err(argv[0], 0,
			"Usage: rdump [-p] "
			"<fs_directory>... <output_directory>");
	ss_set_exit_status(sci_idx, 1);
```
